### PR TITLE
tvp.pl extractor update

### PR DESCRIPTION
**Pull request by d42**
Basically, the public tv overlords have changed the url routing, and some of the api.
As a bonus it handles shows in their own subdomains now :3

### CODE REVIEW

File: youtube_dl/extractor/tvp.py [hunk 1]
```diff
@@ -1,139 +1,247 @@
...
+VIDEO_LISTING_URL = ('http://www.api.v3.tvp.pl/shared/listing.php'
+                     '?dump=json&direct=true&count=-1&parent_id={id}')
+META_URL = 'http://www.tvp.pl/shared/video_data.php?dump=json&video_id={id}'
+TOKENIZER_URL = 'http://www.tvp.pl/shared/cdn/tokenizer_v2.php?object_id={id}'
+FILE_INFO_URL = 'http://www.tvp.pl/pub/stat/videofileinfo?video_id={id}'
+IGNORED_MIMETYPES = 'application/vnd.ms-ss', 'application/x-mpegurl'
 
-from .common import InfoExtractor
+
+class TvpApi:
```
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: I don't see much point extracting it in separate class. Base class inherited from `InfoExtractor` is preferred.

File: youtube_dl/extractor/tvp.py [hunk 2]
```diff
@@ -1,139 +1,247 @@
...
-from __future__ import unicode_literals
+# -*- encoding: utf-8 -*-
+from .common import InfoExtractor, ExtractorError
 
-import re
+VIDEO_LISTING_URL = ('http://www.api.v3.tvp.pl/shared/listing.php'
+                     '?dump=json&direct=true&count=-1&parent_id={id}')
+META_URL = 'http://www.tvp.pl/shared/video_data.php?dump=json&video_id={id}'
+TOKENIZER_URL = 'http://www.tvp.pl/shared/cdn/tokenizer_v2.php?object_id={id}'
+FILE_INFO_URL = 'http://www.tvp.pl/pub/stat/videofileinfo?video_id={id}'
```
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: I don't see much point in there constants since each is only used once.

File: youtube_dl/extractor/tvp.py [hunk 3]
```diff
@@ -1,139 +1,247 @@
...
+        json = self._get_json(VIDEO_LISTING_URL, id)
+        return json
+
+    def meta(self, id):
+        json = self._get_json(META_URL, id)
+        return json
+
+    def info(self, id):
+        json = self._get_json(FILE_INFO_URL, id)
+        return json
```
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: Code duplication in these three methods. Again they are only used once, so there is really no need in them.

File: youtube_dl/extractor/tvp.py [hunk 4]
```diff
@@ -1,139 +1,247 @@
...
+        json = self._get_json(TOKENIZER_URL, id)
+        status = json['status']
+        if status == 'NOT_PLAYABLE':
+            raise ExtractorError("video is not playable")
+        if status != 'OK':
+            raise ExtractorError("unknown status: %s", status)
+        return json['formats']
+
+    def _get_json(self, url, id):
+        id = int(id)
```
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: Avoid shadowing built-in names.

File: youtube_dl/extractor/tvp.py [hunk 5]
```diff
@@ -1,139 +1,247 @@
...
-            m3u8_url = re.sub('([^/]*)\.ism/manifest', r'\1.ism/\1.m3u8', video_url)
-            formats = self._extract_m3u8_formats(m3u8_url, video_id, 'mp4')
+            title = ', '.join([context['website_title'], context['title']])
+        url = context['url']
+        description = context['description_root']
+
+        try:
+            formats = self._format_formats(self.api.formats(id), id)
+        except ExtractorError as e:
+            self.to_screen("%s: %s" % (title, e))
```
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: What's the point of this?

File: youtube_dl/extractor/tvp.py [hunk 6]
```diff
@@ -1,139 +1,247 @@
...
         else:
-            m3u8_url = re.sub('([^/]*)\.ism/manifest', r'\1.ism/\1.m3u8', video_url)
-            formats = self._extract_m3u8_formats(m3u8_url, video_id, 'mp4')
+            title = ', '.join([context['website_title'], context['title']])
+        url = context['url']
+        description = context['description_root']
+
+        try:
+            formats = self._format_formats(self.api.formats(id), id)
+        except ExtractorError as e:
```
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: If any of the formats fails the whole extraction is broken.

File: youtube_dl/extractor/tvp.py [hunk 7]
```diff
@@ -1,139 +1,247 @@
...
+            item_id = ids.pop()
+            listing = self.api.listing(item_id)
+            for item in listing['items']:
+                if 'directory_video' in item['types']:
+                    ids.append(item['_id'])
+                if 'video' in item['types'] and item['is_released']:
+                    meta = self.api.context(item['_id'])
+                    try:
+                        yield self._get_video(meta)
+                    except ExtractorError:
```
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: This defeats the whole point of `ExtractorError`. It should be propagated and handled outside the extractor not muted in extractor.

File: youtube_dl/extractor/tvp.py [hunk 8]
```diff
@@ -1,139 +1,247 @@
...
+
+    def context(self, id):
+        meta = self.meta(id)
+        return meta['context']
+
+    def formats(self, id):
+        json = self._get_json(TOKENIZER_URL, id)
+        status = json['status']
+        if status == 'NOT_PLAYABLE':
+            raise ExtractorError("video is not playable")
```
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: Use single quotes consistently whenever possible. If this is expected error pass `expected=True`.

File: youtube_dl/extractor/tvp.py [hunk 9]
```diff
@@ -1,139 +1,247 @@
...
+        if context['title_root']:
+            title = context['title_root']
+        elif not context['website_title']:
+            title = context['title']
         else:
-            m3u8_url = re.sub('([^/]*)\.ism/manifest', r'\1.ism/\1.m3u8', video_url)
-            formats = self._extract_m3u8_formats(m3u8_url, video_id, 'mp4')
+            title = ', '.join([context['website_title'], context['title']])
+        url = context['url']
+        description = context['description_root']
```
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: Extration should be tolerate to any missing optional data. Anything apart from title and data required for formats extraction considered optional. So if `description_root` is missing if should not break the extraction.

**Comment by dstftw** [comment]
I recommend you looking through the code of any recent extractor for best practices and idiomatic coding style.

File: youtube_dl/extractor/tvp.py (hunk 5 above)
Comment (d42) [inline youtube_dl/extractor/tvp.py]: Sometimes playlists contain videos that don't provide any sources, so I ended up raising exception and ignoring it if it was raised inside a playlist,
Also i wanted to know that something like that happened, so i reraised it here to print out which video was at fault.

Is there a mechanism for handling something like that?

File: youtube_dl/extractor/tvp.py (hunk 9 above)
Comment (d42) [inline youtube_dl/extractor/tvp.py]: API always provides that key, even if value is just an empty string, should I handle it nonetheless?

File: youtube_dl/extractor/tvp.py (hunk 5 above)
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: Video that does not provide any sources sounds like unexpected outcome therefore it should not be muted. If user wants to ignore this she have to pass `-i`.

File: youtube_dl/extractor/tvp.py (hunk 9 above)
Comment (dstftw) [inline youtube_dl/extractor/tvp.py]: Today it provides, tomorrow it doesn't. But extractor is broken across all the packages some of which are not tend to update fast enough.

**Pull request closed without merging**